nsolve:added lambdify printer for uppergamma and lowergamma, also add…

priyank-p0 opened the pull request:
Fixes #12173

priyank-p0 committed:
nsolve:added lambdify printer for uppergamma and lowergamma, also added the unit tests. All tests passed

pspbot7 commented:
@jksuom added the unit tests, earlier the issue was due to some formatting error, all tests ran successfully can you give it a look.

jksuom commented:
There should be tests for both uppergamma and lowergamma. Moreover, the issue was about mpmath, so the lambdify tests should be for mpmath. (`lambdify(..., "mpmath")`, see `test_mpmath_lambda`.)

priyank-p0 committed:
added the test case for lowergamma, and also added the test to the mpmath section.

Earlier the values returned by uppergamma and lowergamma were the same, but now the issue has been solved.

Earlier the values returned by uppergamma and lowergamma were the same, but now the issue has been solved.

Earlier the values returned by uppergamma and lowergamma were the same, but now the issue has been solved.renamed the test file

Revised the unit test as it was giving assertion error with python 3.

Made the requested changes

Made the requested changes

rectified the typo error in uppergamma printer, and removed the extra mpf wrapper

made the required changes, removed two consequitive string operation with one in uppergamma and lowergamma printer.

made requested changes to the unit test

made requested changes to the unit test

made requested changes to the unit test

made requested changes to the unit test

made requested changes to the unit test

made the required changes.

Update test_lambdify.py

Update test_lambdify.py

corrected the formatting of text

pspbot7 commented:
@jksuom added the lowergamma test case, and also transferred the test in the mpmath section

jksuom commented on sympy/utilities/tests/test_lambdify.py:
```diff
@@ -119,6 +119,20 @@ def test_mpmath_lambda():
 @conserve_mpmath_dps
+def test_issue_12173_1():
+    mpmath.mp.dps = 50
+    exp= lambdify((x,y), uppergamma(x,y))(1,2)
```
There should also be a `module` argument "mpmath" in lambdify.

jksuom commented on sympy/utilities/tests/test_lambdify.py:
```diff
@@ -119,6 +119,20 @@ def test_mpmath_lambda():
+
+
+@conserve_mpmath_dps
+def test_issue_12173_2():
```
Both uppergamma and lowergamma can be tested in the same test_issue_12173. There is no need to define another test function.

jksuom commented on sympy/utilities/tests/test_lambdify.py:
```diff
@@ -119,6 +119,20 @@ def test_mpmath_lambda():
+def test_issue_12173_2():
+    mpmath.mp.dps = 50
+    exp= lambdify((x,y), lowergamma(x,y))(1,2)
+    assert exp==mpmath.mpf('0.13533528323661270231781372785917483270168304443359')
```
Is it true that the numerical values of both uppergamma and lowergamma are the same?

pspbot7 commented:
@jksuom I also had the same doubt, tried to search online but couldn't find the answer, will post here if I find the correct answer

pspbot7 commented:
@jksuom the same value issue has been fixed, now the values returned are correct, please take a look

jksuom commented on sympy/utilities/lambdify.py:
```diff
@@ -50,9 +50,9 @@
     "I": "j",
     "ln": "log",
-    #"lowergamma":"lower_gamma",
+    "lowergamma":"gammainc",
```
Are these added lines also used by lambdify?

pspbot7 replied:
yes, they are used to translate the keyword used by sympy to the ones used by mpmath.

jksuom replied:
Since `lowergamma` and `uppergamma` are different functions, I don't think they should have the same translation `gammainc`.

pspbot7 replied:
no, they both have the same keyword in mpmath, but have different parameters i.e. for upper gamma its gammainc(z,a,inf) and for lower gamma its gammainc(z,0,b)

jksuom replied:
Those arguments are set by the mpmath printers, and the current tests are running the printers only. They do not make use of the contents of MPMATH_TRANSLATIONS.

If the translations are added to MPMATH_TRANSLATIONS, it should be tested that the same translation `gammainc` really works for both functions and does not always produce the same values for `uppergamma` and `lowergamma`.

pspbot7 replied:
@jksuom i have fixed the value , now both uppergamma and lowergamma are returning different values

jksuom replied:
How does it affect those values, if the translation lines are are left commented out as they were?

pspbot7 replied:
it doesn't give any error if I comment out the above two lines , so we can remove them

jksuom replied:
I think the lines could be left as they were.

pspbot7 replied:
Ok i will make the changes

pspbot7 commented:
@jksuom all the tests passed when I ran them locally, I don't know why the build failed on Travis CI, can    you rerun the tests

pspbot7 commented:
@jksuom do you have any clue why the following line gives assertion error in python 3 but runs fine in python 3

`   assert exp2==mpmath.mpf('0.86466471676338729768218627214082516729831695556641')`

pspbot7 commented:
@jksuom commented out the two lines

pspbot7 commented:
@jksuom commented out the two lines

pspbot7 commented:
@jksuom all the tests have passed, can you take a look.

pspbot7 commented:
@jksuom do I need to add anything else?

jksuom commented on sympy/printing/lambdarepr.py:
```diff
@@ -348,6 +348,13 @@ def _print_Float(self, e):
         return 'mpf(%s)' % args
 
+    def _print_uppergamma(self,e): #printer for the uppergamma function
+        exp="gammainc({0}, {1}, 6)".format(e.args[0], e.args[1])
```
The third argument looks like `6`. That is probably not right.

jksuom commented on sympy/printing/lambdarepr.py:
```diff
@@ -348,6 +348,13 @@ def _print_Float(self, e):
 
+    def _print_uppergamma(self,e): #printer for the uppergamma function
+        exp="gammainc({0}, {1}, 6)".format(e.args[0], e.args[1])
+        return 'mpf(%s)' %(self._print(exp))
```
This method currently returns `lambda x,y: (mpf(gammainc(x, y, 6)))`. If the value of `gammainc` is of type `mpf`, the extra cast is not necessary.

pspbot7 commented:
@jksuom made the changes, I overlooked the 6 in upper gamma printer, also removed the extra mpf wrapping

jksuom commented on sympy/printing/lambdarepr.py:
```diff
@@ -348,6 +348,13 @@ def _print_Float(self, e):
 
+    def _print_uppergamma(self,e): #printer for the uppergamma function
+        exp="gammainc({0}, {1}, inf)".format(e.args[0], e.args[1])
+        return "(%s)" %(self._print(exp))
```
This combination of two consequitive string formatting operations  (.format and %) could probably be replaced by a single operation with arguments `self._print(e.args[0])` and `self._print(e.args[1])`.

pspbot7 replied:
but self._print() operation is printing the value of the result, while e.args[0],e.args[1] is taking the arguments

jksuom replied:
I mean a single *formatting* operation like in 
`return "gammainc({0}, {1}, inf)".format(self._print(e.args[0]), ...`

Compare with e.g. [this line](https://github.com/sympy/sympy/blob/master/sympy/printing/lambdarepr.py#L198) (though preferably with .format instead of %).

jksuom commented:
The tests should be modified. It is not necessary to calculate the values with 50 decimals to see that correct functions are applied. Hence the lines concerned with dps change could be omitted. In fact, the whole test could be moved to the end of the file where other special printers are tested.

Then the values returned by the lambdified functions should compared with something that has been computed with another method. For example, since `uppergamma(3, 2) == 10*exp(-2)`, the value returned by the lambdified uppergamma should be equal to `10*exp(-2)` within reasonable tolerance.

jksuom commented on sympy/utilities/tests/test_lambdify.py:
```diff
@@ -783,3 +773,13 @@ def test_Indexed():
+def test_issue_12173():
+    exp1= lambdify((x,y), uppergamma(x,y),"mpmath")(1,2)
+    exp2= lambdify((x,y), lowergamma(x,y),"mpmath")(1,2)
+    exp1=simplify(exp1)
```
What is the purpose of `simplify`?

jksuom commented on sympy/utilities/tests/test_lambdify.py:
```diff
@@ -775,11 +774,9 @@ def test_Indexed():
-    exp2=simplify(exp2)
-    ans1=(uppergamma(1,2)).evalf()
-    ans2=(lowergamma(1,2)).evalf()
+    exp1=N(lambdify((x,y), uppergamma(x,y),"mpmath")(1,2))
```
It would be preferable to avoid importing N if that can be avoided. This line could be
```
    exp1 = lambdify((x, y), uppergamma(x, y), "mpmath")(1, 2)
```
(Note the space on either side of = and after each comma.)

jksuom commented on sympy/utilities/tests/test_lambdify.py:
```diff
@@ -775,11 +774,9 @@ def test_Indexed():
+    exp2= N(lambdify((x,y), lowergamma(x,y),"mpmath")(1,2))
+    ans1=N((uppergamma(1,2)))
+    ans2=N((lowergamma(1,2)))
     assert exp1==ans1
```
If the automatic comparisons work as expected, this line could be
```
    assert exp1 == uppergamma(1, 2).evalf()
```
The difference between N and evalf is that N will begin by sympifying its argument. In this case we know that `uppergamma(1, 2)` is a SymPy object. Hence it is preferable to use evalf.

(The mpf object exp1 on the left hand side can be directly compared with the SymPy Float on the right hand side. Hence N is not necessary.)

pspbot7 replied:
made the changes, can you take a look

pspbot7 commented:
@jksuom can you review it?

jksuom commented on sympy/utilities/tests/test_lambdify.py:
```diff
@@ -771,3 +772,10 @@ def test_Indexed():
+
+def test_issue_12173():
+    #test for issue 12173
+    exp1=lambdify((x,y), uppergamma(x,y),"mpmath")(1,2)
```
SymPy code formatting expects that `=` (and `==`) have a space on either side. There should also be a space after each comma.

jksuom commented:
The printers look good. Only the formatting of the tests should be fixed.

jksuom commented:
Now what remains to be done is to squash the series of commits into a single one. (This may help: https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History, Squashing Commits. After rebasing, the force flag has to be added to make the push command work: `git push -f github`.)

pspbot7 commented:
@jksuom squashed the commits , can you take a look

jksuom commented:
Thanks! This is ready.
